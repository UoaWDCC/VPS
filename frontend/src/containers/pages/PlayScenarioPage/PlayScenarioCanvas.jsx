import CountdownTimer from "../../../components/TimerComponent";
import componentResolver from "./componentResolver";

/**
 * This component displays the scene components on the screen when playing a scenario
 *
 * @component
 */

export default function PlayScenarioCanvas({
  scene,
  incrementor,
  reset,
  setAddFlags,
  setRemoveFlags,
}) {
  return (
    <>
      {scene.components?.map((component, index) => {
        let action = () =>
          component.nextScene && incrementor(component.nextScene);
        switch (component.type) {
          case "RESET_BUTTON":
            action = reset;
            break;
          case "BUTTON":
            setAddFlags(component.flagAdditions);
            setRemoveFlags(component.flagDeletions);
            break;
          default:
            break;
        }

        return componentResolver(component, index, action);
      })}
      {scene.time && (
        <CountdownTimer
          targetDate={new Date().setSeconds(
            new Date().getSeconds() + scene.time
          )}
          sceneTime={scene.time}
        />
      )}
    </>
  );
}
