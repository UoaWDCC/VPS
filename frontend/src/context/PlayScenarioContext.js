import React from "react";

const PlayScenarioContext = React.createContext();

export default PlayScenarioContext;
