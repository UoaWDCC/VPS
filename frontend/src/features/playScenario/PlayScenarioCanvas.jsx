import { useState } from "react";
import ResetConfirmationModal from "./modals/ResetConfirmationModal";
import { buildVisualScene } from "../authoring/pipeline";
import TextBox from "../authoring/elements/TextBox";
import Speech from "../authoring/elements/Speech";
import Ellipse from "../authoring/elements/Ellipse";
import Box from "../authoring/elements/Box";
import Image from "../authoring/elements/Image";
import Line from "../authoring/elements/Line";

const componentMap = {
  textbox: TextBox,
  speech: Speech,
  ellipse: Ellipse,
  box: Box,
  image: Image,
  line: Line,
};

function resolve(component) {
  const Fc = componentMap[component.type];
  if (Fc) return <Fc key={component.id} {...component} />;
  return null;
}

function injectStateVariables(scene, stateVariables) {
  if (!stateVariables) return scene;

  const varMap = new Map(stateVariables.map((v) => [v.name, v.value]));
  const regex = /\$\$(.*?)\$\$/g;

  return {
    ...scene,
    components: Object.fromEntries(
      Object.entries(scene.components).map(([id, component]) => {
        if (component.type !== "textbox") return [id, component];

        const newBlocks = component.document.blocks.map((block) => {
          const newSpans = [];

          for (const span of block.spans) {
            const pieces = [];
            let lastIndex = 0;
            let match;

            while ((match = regex.exec(span.text)) !== null) {
              if (match.index > lastIndex) {
                pieces.push(span.text.slice(lastIndex, match.index));
              }
              const variableName = match[1];
              if (varMap.has(variableName)) {
                pieces.push(String(varMap.get(variableName)));
              } else {
                pieces.push(match[0]);
              }
              lastIndex = regex.lastIndex;
            }

            if (lastIndex < span.text.length) {
              pieces.push(span.text.slice(lastIndex));
            }

            regex.lastIndex = 0;

            // Only add non-empty pieces
            for (const text of pieces) {
              if (text !== "") {
                newSpans.push({ text });
              }
            }

            // Fallback: if substitution produced no visible text, preserve the original
            if (pieces.length === 0) {
              newSpans.push(span);
            }
          }

          return { ...block, spans: newSpans };
        });

        return [
          id,
          {
            ...component,
            document: { ...component.document, blocks: newBlocks },
          },
        ];
      })
    ),
  };
}

/**
 * This component displays the scene components on the screen when playing a scenario
 *
 * @component
 */

export default function PlayScenarioCanvas({
  scene,
  buttonPressed,
  reset,
  setAddFlags,
  setRemoveFlags,
  stateVariables,
}) {
  // const [isModalOpen, setIsModalOpen] = useState(false);

  // const handleResetClick = () => {
  //   setIsModalOpen(true);
  // };

  // const handleConfirmReset = async () => {
  //   setIsModalOpen(false);
  //   try {
  //     await reset();
  //   } catch (error) {
  //     console.error("Error during reset confirmation:", error);
  //   }
  // };

  // const handleCancelReset = () => {
  //   setIsModalOpen(false);
  // };

  const sceneToRender = injectStateVariables(scene, stateVariables);

  const components = Object.values(buildVisualScene(sceneToRender).components)
    .sort((a, b) => a.zIndex - b.zIndex)
    .map((c) => {
      const resolved = resolve(c);
      if (c.clickable) {
        return (
          <g
            key={c.id}
            className="cursor-pointer"
            onClick={() => buttonPressed(c)}
          >
            {resolved}
          </g>
        );
      }
      return resolved;
    });

  return (
    <div className="bg-white" style={{ width: "100vw", height: "100vh" }}>
      <svg id="main" className="w-full h-full" viewBox="0 0 1920 1080">
        <rect x="0" y="0" width="1920" height="1080" fill="white" />
        {components}
      </svg>
      {/* <ResetConfirmationModal
        isOpen={isModalOpen}
        onConfirm={handleConfirmReset}
        onClose={handleCancelReset}
      /> */}
    </div>
  );
}
