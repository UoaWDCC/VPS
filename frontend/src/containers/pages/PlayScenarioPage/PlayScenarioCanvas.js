import React, { useState, useContext } from "react";
import { useGet, usePost, usePut } from "../../../hooks/crudHooks";
import componentResolver from "./componentResolver";
import PlayScenarioContext from "../../../context/PlayScenarioContext";
import ProgressBar from "./progressBar";
import CountdownTimer from "../../../components/TimerComponent";
import AuthenticationContext from "../../../context/AuthenticationContext";

/**
 * This component displays the scene components on the screen when playing a scenario
 *
 * @component
 */
export default function PlayScenarioCanvas(props) {
  const { progress, graph } = props;
  const [currentScene, setCurrentScene] = useState(null);
  const [maxProgress, setMaxProgress] = useState(0);
  // eslint-disable-next-line no-unused-vars
  const { scenarioId, currentSceneId, setCurrentSceneId } =
    useContext(PlayScenarioContext);

  const { user } = useContext(AuthenticationContext);

  useGet(
    `/api/scenario/${scenarioId}/scene/full/${currentSceneId}`,
    setCurrentScene,
    false
  );

  const componentOnClick = (component) => {
    if (component.type === "BUTTON" && component.nextScene !== "") {
      setMaxProgress(Math.max(progress, maxProgress));
      graph.visit(component.nextScene);
      if (graph.isEndScene(component.nextScene)) {
        // update db

        const userData = {
          name: user.displayName,
          uid: user.uid,
          email: user.email,
        };

        usePost("/api/user", userData);
      }
      setCurrentSceneId(component.nextScene);
    }
  };

  return (
    <>
      {currentScene?.components?.map((component, index) =>
        componentResolver(component, index, () => componentOnClick(component))
      )}
      <ProgressBar value={Math.max(progress, maxProgress) * 100} />
      {currentScene?.time ? (
        <CountdownTimer
          targetDate={new Date().setSeconds(
            new Date().getSeconds() + currentScene?.time
          )}
        />
      ) : null}
    </>
  );
}
