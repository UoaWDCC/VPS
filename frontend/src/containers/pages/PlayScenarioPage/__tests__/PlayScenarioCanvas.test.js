import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import PlayScenarioCanvas from "../PlayScenarioCanvas";
import PlayScenarioContext from "../../../../context/PlayScenarioContext";

test("PlayScenarioCanvas component snapshot test", () => {
  const context = {
    scenarioId: "scenarioID",
    currentSceneId: "scenarioID",
    setCurrentSceneId: () => {},
  };

  const { baseElement } = render(
    <BrowserRouter>
      <PlayScenarioContext.Provider value={context}>
        <PlayScenarioCanvas />
      </PlayScenarioContext.Provider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
