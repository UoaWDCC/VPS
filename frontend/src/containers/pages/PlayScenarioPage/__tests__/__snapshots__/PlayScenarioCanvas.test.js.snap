// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayScenarioCanvas component snapshot test 1`] = `
<body>
  <div />
</body>
`;
